Bio/PDB/__init__.py

```diff
@@ -88,3 +97,67 @@
     from .SASA import ShrakeRupley
 except ImportError:
     pass
+
+_FormatToParser = {"pdb": PDBParser(), "cif": MMCIFParser()}
+
+
+def read(handle, format, id=None):
+    """Parse a file or handle using one of the available parsers to generate a structure.
+
+    Arguments:
+     - handle    - handle to the file, or the filename as a string
+     - format    - string describing the file format. can be [pdb, cif, mmtf]
+     - id        - optional id of the structure, if None header information or then base name is used
+
+    If you have the file name in a string 'filename', use:
+
+    >>> from Bio import PDB
+    >>> filename = "1A8O.pdb"
+    >>> PDB.read(filename, format="pdb")
+
+    The function also supports file handles and gzip files
+
+    >>> from Bio import PDB
+    >>> PDB.read("1A8O.pdb.gz", format="pdb")
+    """
+    # TODO *kwargs for the parser to allow QUIET or permissive mode
+
+    # Try and give helpful error messages:
+    if not isinstance(format, str):
+        raise TypeError("Need a string for the file format (lower case)")
+    if not format:
+        raise ValueError("Format required (lower case string)")
+    if format != format.lower():
+        raise ValueError("Format string '%s' should be lower case" % format)
+
+    # need to use different parser for mmtf versus pdb,cif
+    if format == "mmtf":
+        # TODO: Raise warning if ID is set. MMTFParser does not support passing an id
+        # MMTF needs a seperate parser as get_structure does only take 1 positional argument and does not support file handles
+        if isinstance(handle, str):
+            parser = MMTFParser()
+            return parser.get_structure(handle)
+        else:
+            raise TypeError("'%s' parser only accepts str not handle" % format)
+
+    elif format in _FormatToParser:  # Map the file format to a parser
+        # if we are provided a path to a file, check if gzipped,
+        # else open handle directly
+        if isinstance(handle, str):
+            encoding = guess_type(handle)[1]  # uses file extension
```

Comment:
Here, you could uses `guess_type` in a `try: else:` block. This allows you to accept `str`, `bytes`, and `os.PathLike` objects as valid file names.

Reply:
I thought `guess_type` was actually checking the file, not just the extension. If that's the case, we might as well ignore `mimetypes.guess_type` and use `Path.suffix` to match the dictionary mapping extensions to parsers (`_FormatToParser`).

Reply:
I know that e.g biological assemblies come in pdb1.gz format so then we would need to match just the gz part because a file might be of format pdb but have a different extension. 
`guess_type` really only looks at the filename but I though it'd be nice and readable this way. 
Alternatively one could look at the first two bytes of the file and determine if it is  gzipped but afaik there are also very few edge cases where this does not work 100% reliably. I think `gunzip` also only looks at the filename whether it is .gz extension so I think this is reasonably safe this way. 

Beginning Python3.8 there is also a gzip.BadGzipFile Exception but I guess for backwards compatibility this can only be implemented in the somewhat clunky wat: 
```
import gzip
with gzip.open(filename, 'r') as fh:
    try:
        fh.read(1)
    except OSError:
        print('input_file is not a valid gzip file by OSError')
    except gzip.BadGzipFile: # only Python 3.8+
        print('input_file is not a valid gzip file by BadGzipFile')
```